test/lib/tpunit++.cpp

```diff
@@ -389,28 +391,43 @@ void tpunit::TestFixture::tpunit_detail_do_methods(tpunit::TestFixture::method*
 
 void tpunit::TestFixture::tpunit_detail_do_tests(TestFixture* f) {
     method* t = f->_tests;
-    recursive_mutex& m = *(f->_mutex);
+    list<thread> testThreads;
     while(t) {
-       int _prev_assertions = f->_stats._assertions;
-       int _prev_exceptions = f->_stats._exceptions;
-       f->testOutputBuffer = "";
-       tpunit_detail_do_methods(f->_befores);
-       tpunit_detail_do_method(t);
-       tpunit_detail_do_methods(f->_afters);
-       if(_prev_assertions == f->_stats._assertions && _prev_exceptions == f->_stats._exceptions) {
-          lock_guard<recursive_mutex> lock(m);
-          printf("\xE2\x9C\x85 %s\n", t->_name);
-          tpunit_detail_stats()._passes++;
-       } else {
-          lock_guard<recursive_mutex> lock(m);
-
-          // Dump the test buffer if the test included any log lines.
-          f->printTestBuffer();
-          printf("\xE2\x9D\x8C !FAILED! \xE2\x9D\x8C %s\n", t->_name);
-          tpunit_detail_stats()._failures++;
-          tpunit_detail_stats()._failureNames.emplace(t->_name);
-       }
-       t = t->_next;
+        testThreads.push_back(thread([t, f]() {
+            recursive_mutex& m = *(f->_mutex);
+            f->_stats._assertions = 0;
+            f->_stats._exceptions = 0;
+            f->testOutputBuffer = "";
+            tpunit_detail_do_methods(f->_befores);
+            tpunit_detail_do_method(t);
+            tpunit_detail_do_methods(f->_afters);
+
+            // No new assertions or exceptions. This not currently synchronized correctly. They can cause tests that
+            // passed to appear failed when another test failed while this test was running. They cannot cause failed
+            // tests to appear to have passed.
+            if(!f->_stats._assertions && !f->_stats._exceptions) {
+                lock_guard<recursive_mutex> lock(m);
```

Comment:
I'm not sure I follow how `recursive_mutex` works with `lock_guard`. based on https://www.cplusplus.com/reference/mutex/recursive_mutex/ it seems like you'd need to call `unlock` for every time you called `recursive_mutex` - but maybe this is my misunderstanding?  `lock_guard` is meant to ensure "the duration of the mutex object shall extend at least until the destruction of the lock_guard that locks it" https://www.cplusplus.com/reference/mutex/lock_guard/?kw=lock_guard 

I also came across this which doesn't describe needing to do what I'm confused on  http://www.gerald-fahrnholz.eu/sw/online_doc_multithreading/html/group___grp_mutexes_and_locks.html so I'm not too concerned about it.

Reply:
`recursive_mutex` is just a `mutex` that you can lock multiple times in the same thread.

Normal mutex:
```
mutex m;
{
    lock_guard l1(m);
    {
        lock_guard l2(m); // Never locks, because it's waiting for m1 to unlock first.
    }
}
```

Recursive mutex:
```
mutex m;
{
    lock_guard l1(m);
    {
        lock_guard l2(m); // This thread owns the lock l1 so we can lock again.
    } // l2 goes out of scope and the lock is released.
} // l1 goes out of scope and the lock is released.
```